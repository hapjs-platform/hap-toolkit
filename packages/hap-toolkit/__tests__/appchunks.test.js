/*
 * Copyright (c) 2021-present, the hapjs-platform Project Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

const path = require('path')
const del = require('del')
const { copyApp, wipeDynamic } = require('hap-dev-utils')
const { compile } = require('../lib')

function testContentMatch(stats, projectRoot) {
  projectRoot = path.resolve(projectRoot)
  const projectRootReg = new RegExp(projectRoot, 'g')
  const json = stats.toJson({ source: true })
  json.modules
    .filter((mod) => mod.source)
    .forEach((module) => {
      expect(wipeDynamic(module.source, [[projectRootReg, '<project-root>']])).toMatchSnapshot()
    })
}

function getJson(stats) {
  return stats
    .toJson()
    .assets.filter((_) => _.name.endsWith('.json'))
    .map((_) => _.name)
}

describe('split common component on a project', () => {
  const platform = 'native'
  let projectRoot
  let buildDir
  const expectResult = ['app-chunks.json']

  beforeAll(async () => {
    const testAppDir = path.resolve(__dirname, '../fixtures/commonComponent-app')
    projectRoot = await copyApp(testAppDir)
    buildDir = path.resolve(projectRoot, 'build')
  })
  // build 模式添加 --split-chunks-mode=smart 参数时，需要把公共组件抽取出来
  it('build results include split chunks for [SMART] ', async () => {
    const mode = 'dev'
    const { stats } = await compile(platform, mode, false, {
      cwd: projectRoot,
      splitChunksMode: 'SMART',
      enableLazyComponent: true
    })
    expect(stats.hasErrors()).toBe(false)
    const result = getJson(stats)

    expect(result).toEqual(expect.arrayContaining(expectResult))
    testContentMatch(stats, projectRoot)
  }, 50000)

  afterAll(async () => {
    await del([projectRoot])
  })
})
