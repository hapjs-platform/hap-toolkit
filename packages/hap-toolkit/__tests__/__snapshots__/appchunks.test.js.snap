// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`split common component on a project build results include split chunks for [SMART]  1`] = `
"require(\\"!../../../../hap-dsl-xvm/lib/loaders/ux-loader.js?cwd=<project-root>&type=import!./component/part1.ux?uxType=comp&name=part1\\")
require(\\"!../../../../hap-dsl-xvm/lib/loaders/ux-loader.js?cwd=<project-root>&type=import!./component/part2.ux?uxType=comp&name=part2\\")


var $app_style$ = {}
var $app_script$ = require(\\"!../../../../hap-dsl-xvm/lib/loaders/script-loader.js!../../../../../node_modules/@hap-toolkit/packager/lib/loaders/module-loader.js!../../../../../node_modules/@hap-toolkit/packager/lib/loaders/manifest-loader.js?path=<project-root>/src!../../../../../node_modules/babel-loader/lib/index.js?cwd=<project-root>&cacheDirectory&comments=false&configFile=<project-root>/babel.config.js!../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=script!./app.ux?uxType=app\\")

$app_define$('@app-application/app', [], function ($app_require$, $app_exports$, $app_module$) {
  
  $app_script$($app_module$, $app_exports$, $app_require$)
  if ($app_exports$.__esModule && $app_exports$.default) {
    $app_module$.exports = $app_exports$.default
  }
  $app_module$.exports.manifest = require(\\"./manifest.json\\")
  $app_module$.exports.style = { list: [ $app_style$ ] }
  
})
$app_bootstrap$('@app-application/app', { packagerVersion: QUICKAPP_TOOLKIT_VERSION })
"
`;

exports[`split common component on a project build results include split chunks for [SMART]  2`] = `
"
var $app_style$ = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/style-loader.js?index=0&type=style!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=style!./index.ux?uxType=page\\")

var $app_script$ = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/script-loader.js!../../../../../../node_modules/@hap-toolkit/packager/lib/loaders/module-loader.js!../../../../../../node_modules/babel-loader/lib/index.js?cwd=<project-root>&cacheDirectory&plugins[]=<CWD>/packages/hap-dsl-xvm/lib/loaders/babel-plugin-jsx.js&comments=false&configFile=<project-root>/babel.config.js!../../../../../hap-dsl-xvm/lib/loaders/access-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=script!./index.ux?uxType=page\\")

$app_define$('@app-component/index', [], function($app_require$, $app_exports$, $app_module$) {
     $app_script$($app_module$, $app_exports$, $app_require$)
         if ($app_exports$.__esModule && $app_exports$.default) {
            $app_module$.exports = $app_exports$.default
        }
     $app_module$.exports.template = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/template-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=template!./index.ux?uxType=page&\\")

     $app_module$.exports.style = $app_style$
})
$app_bootstrap$('@app-component/index',{ packagerVersion: QUICKAPP_TOOLKIT_VERSION })"
`;

exports[`split common component on a project build results include split chunks for [SMART]  3`] = `
"
var $app_style$ = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/style-loader.js?index=0&type=style!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=style!./index.ux?uxType=card\\")

var $app_script$ = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/script-loader.js!../../../../../../node_modules/@hap-toolkit/packager/lib/loaders/module-loader.js!../../../../../../node_modules/babel-loader/lib/index.js?cwd=<project-root>&cacheDirectory&plugins[]=<CWD>/packages/hap-dsl-xvm/lib/loaders/babel-plugin-jsx.js&comments=false&configFile=<project-root>/babel.config.js!../../../../../hap-dsl-xvm/lib/loaders/access-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=script!./index.ux?uxType=card\\")

$app_define$('@app-component/index', [], function($app_require$, $app_exports$, $app_module$) {
     $app_script$($app_module$, $app_exports$, $app_require$)
         if ($app_exports$.__esModule && $app_exports$.default) {
            $app_module$.exports = $app_exports$.default
        }
     $app_module$.exports.template = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/template-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=template!./index.ux?uxType=card&\\")

     $app_module$.exports.style = $app_style$
})
$app_bootstrap$('@app-component/index',{ packagerVersion: QUICKAPP_TOOLKIT_VERSION })"
`;

exports[`split common component on a project build results include split chunks for [SMART]  4`] = `
"$app_define_wrap$('@app-component/part1', function () {

var $app_style$ = {}
var $app_script$ = null
$app_define$('@app-component/part1', [], function($app_require$, $app_exports$, $app_module$) {
     $app_module$.exports.template = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/template-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=template!./part1.ux?uxType=comp&\\")

})
})"
`;

exports[`split common component on a project build results include split chunks for [SMART]  5`] = `
"$app_define_wrap$('@app-component/part2', function () {

var $app_style$ = {}
var $app_script$ = null
$app_define$('@app-component/part2', [], function($app_require$, $app_exports$, $app_module$) {
     $app_module$.exports.template = require(\\"!../../../../../hap-dsl-xvm/lib/loaders/template-loader.js!../../../../../hap-dsl-xvm/lib/loaders/fragment-loader.js?index=0&type=template!./part2.ux?uxType=comp&\\")

})
})"
`;

exports[`split common component on a project build results include split chunks for [SMART]  6`] = `
"module.exports = function __scriptModule__ (module, exports, $app_require$){\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _default = {};
exports.default = _default;}"
`;

exports[`split common component on a project build results include split chunks for [SMART]  7`] = `
"{
  \\"package\\": \\"com.application.demo\\",
  \\"name\\": \\"app\\",
  \\"versionName\\": \\"1.0.0\\",
  \\"versionCode\\": \\"1\\",
  \\"minPlatformVersion\\": 1070,
  \\"icon\\": \\"/Common/logo.png\\",
  \\"features\\": [],
  \\"permissions\\": [{ \\"origin\\": \\"*\\" }],
  \\"config\\": {
    \\"logLevel\\": \\"off\\"
  },
  \\"router\\": {
    \\"entry\\": \\"Demo\\",
    \\"pages\\": {
      \\"Demo\\": {
        \\"component\\": \\"index\\"
      }
    },
    \\"widgets\\": {
      \\"CardDemo\\": {
        \\"name\\": \\"CardDemo\\",
        \\"decription\\": \\"快应用卡片展示\\",
        \\"component\\": \\"index\\",
        \\"path\\": \\"/CardDemo\\",
        \\"features\\": []
      }
    }
  }
}
"
`;

exports[`split common component on a project build results include split chunks for [SMART]  8`] = `
"module.exports = {
  \\".demo-page\\": {
    \\"flexDirection\\": \\"column\\",
    \\"justifyContent\\": \\"center\\",
    \\"alignItems\\": \\"center\\"
  },
  \\".title\\": {
    \\"fontSize\\": \\"40px\\",
    \\"textAlign\\": \\"center\\"
  },
  \\".btn\\": {
    \\"width\\": \\"550px\\",
    \\"height\\": \\"86px\\",
    \\"marginTop\\": \\"75px\\",
    \\"borderRadius\\": \\"43px\\",
    \\"backgroundColor\\": \\"#09ba07\\",
    \\"fontSize\\": \\"30px\\",
    \\"color\\": \\"#ffffff\\"
  }
}"
`;

exports[`split common component on a project build results include split chunks for [SMART]  9`] = `
"module.exports = function __scriptModule__ (module, exports, $app_require$){\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _default = {
  private: {
    title: '示例页面'
  }
};
exports.default = _default;
const moduleOwn = exports.default || module.exports;
const accessors = ['public', 'protected', 'private'];
if (moduleOwn.data && accessors.some(function (acc) {
  return moduleOwn[acc];
})) {
  throw new Error('页面VM对象中的属性data不可与\\"' + accessors.join(',') + '\\"同时存在，请使用private替换data名称');
} else if (!moduleOwn.data) {
  moduleOwn.data = {};
  moduleOwn._descriptor = {};
  accessors.forEach(function (acc) {
    const accType = typeof moduleOwn[acc];
    if (accType === 'object') {
      moduleOwn.data = Object.assign(moduleOwn.data, moduleOwn[acc]);
      for (const name in moduleOwn[acc]) {
        moduleOwn._descriptor[name] = {
          access: acc
        };
      }
    } else if (accType === 'function') {
      console.warn('页面VM对象中的属性' + acc + '的值不能是函数，请使用对象');
    }
  });
}}"
`;

exports[`split common component on a project build results include split chunks for [SMART]  10`] = `
"module.exports = {
  \\"type\\": \\"div\\",
  \\"attr\\": {},
  \\"classList\\": [
    \\"demo-page\\"
  ],
  \\"children\\": [
    {
      \\"type\\": \\"text\\",
      \\"attr\\": {
        \\"value\\": function () {return '' + '欢迎打开' + (this.title)}
      },
      \\"classList\\": [
        \\"title\\"
      ]
    },
    {
      \\"type\\": \\"part1\\",
      \\"attr\\": {}
    },
    {
      \\"type\\": \\"part2\\",
      \\"attr\\": {}
    }
  ]
}"
`;

exports[`split common component on a project build results include split chunks for [SMART]  11`] = `
"module.exports = {
  \\".demo-page\\": {
    \\"flex\\": 1,
    \\"flexDirection\\": \\"column\\",
    \\"alignItems\\": \\"center\\"
  }
}"
`;

exports[`split common component on a project build results include split chunks for [SMART]  12`] = `
"module.exports = function __scriptModule__ (module, exports, $app_require$){\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _default = {
  private: {
    text: '欢迎使用快应用卡片',
    des: '卡片是一种轻量级的快应用，快应用卡片使快应用能够在其他app和系统里提供可扩展的app功能'
  }
};
exports.default = _default;
const moduleOwn = exports.default || module.exports;
const accessors = ['public', 'protected', 'private'];
if (moduleOwn.data && accessors.some(function (acc) {
  return moduleOwn[acc];
})) {
  throw new Error('页面VM对象中的属性data不可与\\"' + accessors.join(',') + '\\"同时存在，请使用private替换data名称');
} else if (!moduleOwn.data) {
  moduleOwn.data = {};
  moduleOwn._descriptor = {};
  accessors.forEach(function (acc) {
    const accType = typeof moduleOwn[acc];
    if (accType === 'object') {
      moduleOwn.data = Object.assign(moduleOwn.data, moduleOwn[acc]);
      for (const name in moduleOwn[acc]) {
        moduleOwn._descriptor[name] = {
          access: acc
        };
      }
    } else if (accType === 'function') {
      console.warn('页面VM对象中的属性' + acc + '的值不能是函数，请使用对象');
    }
  });
}}"
`;

exports[`split common component on a project build results include split chunks for [SMART]  13`] = `
"module.exports = {
  \\"type\\": \\"div\\",
  \\"attr\\": {},
  \\"classList\\": [
    \\"demo-page\\"
  ],
  \\"children\\": [
    {
      \\"type\\": \\"text\\",
      \\"attr\\": {
        \\"value\\": function () {return this.text}
      }
    },
    {
      \\"type\\": \\"text\\",
      \\"attr\\": {
        \\"value\\": function () {return this.des}
      }
    }
  ]
}"
`;

exports[`split common component on a project build results include split chunks for [SMART]  14`] = `
"module.exports = {
  \\"type\\": \\"div\\",
  \\"attr\\": {},
  \\"children\\": [
    {
      \\"type\\": \\"text\\",
      \\"attr\\": {
        \\"value\\": \\"part1\\"
      }
    }
  ]
}"
`;

exports[`split common component on a project build results include split chunks for [SMART]  15`] = `
"module.exports = {
  \\"type\\": \\"div\\",
  \\"attr\\": {},
  \\"children\\": [
    {
      \\"type\\": \\"text\\",
      \\"attr\\": {
        \\"value\\": \\"part2\\"
      }
    }
  ]
}"
`;
